### DISCUSSION TITLE
micropython uart.read() in SAMD51 is incomplete (only 2 bytes returned)

### DISCUSSION THREAD

**Discussion by cajomferro** [micropython/micropython#10018] [Hardware & Peripherals] [answered]
I have an ATSAMD51 microcontroller integrated in a project-specific board which has also a xbee3 RF module. This xbee3 module is configured for Transparent Mode and the device type is 'router'. I can check this on the XCTU software.

Now, my goal was to install micropython on ATSAMD51 and then send commands to the xbee3 via UART since unfortunately I am not aware of any micropython library that can interact with the xbee. And just out of confusion, I don't want to use the xbee with micropython.

```
++++++++++++++           ++++++++++++++
++ ATSAMD51 ++ --UART--> ++ XBEE3 RF ++
++++++++++++++           ++++++++++++++
(running                 (transparent
micropython)             mode)
```

I successfully installed the micropython on the ATSAMD51 and I managed to send commands to the xbee3 via UART. I know this works since when I do `uart.write('1234')` I can check on the XCTU console that the coordinator node receives this message.
However, If I try to send xbee3 AT commands, the `uart.read()` only gets 2 bytes at most.

Example:
`screen /dev/tty.usbmodem0000000000001 115200` # enter the python REPL of the ATSAMD51

```
>>> from machine import UART, Pin
>>> uart = UART(5, rx=Pin('PB02'), tx=Pin('PB03'), baudrate=115200)
>>> uart
UART(5, baudrate=115200, bits=8, parity=None, stop=1, timeout=1, timeout_char=1, rxbuf=256, txbuf=256)
>>> uart.write('1234') # this actually transmits the message to the coordinator node
4
>>> uart.write('+++') # entering in command mode
3
>>> uart.read()
b'OK\r'
>>> uart.write('ATID\r')
5
>>> uart.read()
b'FO\r' # this should be FOO which is the device PAN ID
>>> uart.write('ATXX\r')
5
>>> uart.read()
b'ER\r' # this should be ERROR
```

Any command I send I always get only 2 characters + the \\r (carriage return). I've tried different baudrates and timeouts and If I try to connect to the xbee directly from the computer via serial this works as expected. So I guess this is somehow related with the micropython UART configuration in the SAMD51?

I'm thinking that maybe the fact that the xbee returns a `\r` might mess up with the micropython's UART implementation? Perhaps the xbee sends chunks of two characters? I am just guessing here and run out of solutions..

As for the `\r` I think that is actually part of the xbee protocol (at least judging for [this link](https://www.digi.com/support/forum/4054/xbee-control-through-serial-port-uart): "The XBee expects just a carriage return character to end a command, and will similarly terminate strings it sends with a carriage return.")

Thank you!

**Comment by robert-hh** [comment]
Thank you for your report. I do not have an XBEE modem here to replicate the situation, but I can replicate it by simply connecting a second terminal. The problem is caused by the write clearing the Receive interrupt enable bit. I changed the code, but have to test it a little bit. Which board/firmware do you use? Or do you build the firmware yourself?

~~There seems to be a timing problem in the receiver, which I have to check. You can avoid that to set the timeout to a small value, like 2.
uart = UART(5, rx=Pin('PB02'), tx=Pin('PB03'), baudrate=115200, timeout=10)~~

**Reply by cajomferro** [reply]
Nice! Thank you!! I built the firmware myself based on the SPARKFUN_SAMD51_THING_PLUS port.

**Comment by robert-hh** [comment, marked as answer]
So I made a PR #10021, which fixed that behavior. I do not know how long it takes until that change is in the main line. Until then, you can build your code based on that PR, or I can provide a firmware image for you.

**Reply by cajomferro** [reply]
Oh yeah! That fixed it!! Awesome!! Thank you @robert-hh I was getting crazy with this ^^

**Comment by robert-hh** [comment]
That's good. Sorry for the confusion.
